# Ejercicios Pair Programming - Modulo 3 - Sprint 1
## Lección 7 - ANOVA
### Pilar García Murillo y Sonia Ruiz Pérez

En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.

Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.

📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [5]:
# Importamos las librerías necesarias.
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm
from statsmodels.formula.api import ols
pd.options.display.max_columns=None   #Para que se muestren todas las columnas de los dataframes.

In [6]:
#Creamos un dataframe con el csv de los datos de nuestro estudio.
df = pd.read_csv("files/taxis3_estandar.csv", index_col=0)
df.head(3)

pickup              dropoff  passengers  distance  fare   tip  \
0  2019-03-23 20:21:09  2019-03-23 20:27:24           1      1.60   7.0  2.15   
1  2019-03-04 16:11:55  2019-03-04 16:19:00           1      0.79   5.0  0.00   
2  2019-03-27 17:53:01  2019-03-27 18:00:25           1      1.37   7.5  2.36   

   tolls  total   color      payment            pickup_zone  \
0    0.0  12.95  yellow  credit card        Lenox Hill West   
1    0.0   9.30  yellow         cash  Upper West Side South   
2    0.0  14.16  yellow  credit card          Alphabet City   

            dropoff_zone pickup_borough dropoff_borough  tip_no_zeros  \
0    UN/Turtle Bay South      Manhattan       Manhattan  2.150000e+00   
1  Upper West Side South      Manhattan       Manhattan  1.000000e-10   
2           West Village      Manhattan       Manhattan  2.360000e+00   

     tip_box  passengers_robust  distance_robust  fare_robust  tolls_robust  \
0   0.787753                0.0        -0.022624    -0.294118           0.0   
1 -10.996744                0.0        -0.389140    -0.529412           0.0   
2   0.886769                0.0        -0.126697    -0.235294           0.0   

   total_robust  
0     -0.127368  
1     -0.511579  
2      0.000000

In [11]:
# Comenzamos con nuestro análisis ANOVA
lm = ols("""tip_box ~ color + payment + pickup_borough + dropoff_borough + passengers_robust + 
   distance_robust + fare_robust + tolls_robust + total_robust""", data=df).fit()
sm.stats.anova_lm(lm)

df         sum_sq        mean_sq             F  \
color                 1.0   18272.648805   18272.648805   4352.526472   
payment               1.0  139228.882164  139228.882164  33164.178972   
pickup_borough        3.0    4501.882088    1500.627363    357.447921   
dropoff_borough       4.0     340.016917      85.004229     20.247921   
passengers_robust     1.0      24.073400      24.073400      5.734260   
distance_robust       1.0     173.567475     173.567475     41.343597   
fare_robust           1.0      98.964969      98.964969     23.573356   
tolls_robust          1.0     154.670350     154.670350     36.842321   
total_robust          1.0   19793.136892   19793.136892   4714.705197   
Residual           6326.0   26557.627412       4.198171           NaN   

                          PR(>F)  
color               0.000000e+00  
payment             0.000000e+00  
pickup_borough     1.917500e-214  
dropoff_borough     1.371275e-16  
passengers_robust   1.666600e-02  
distance_robust     1.370135e-10  
fare_robust         1.231293e-06  
tolls_robust        1.354981e-09  
total_robust        0.000000e+00  
Residual                     NaN

In [12]:
#Obtenemos un resumen de la recta que ajusta a neustro modelo.
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_box   R-squared:                       0.873
Model:                            OLS   Adj. R-squared:                  0.873
Method:                 Least Squares   F-statistic:                     3107.
Date:                Mon, 23 Jan 2023   Prob (F-statistic):               0.00
Time:                        09:53:38   Log-Likelihood:                -13539.
No. Observations:                6341   AIC:                         2.711e+04
Df Residuals:                    6326   BIC:                         2.721e+04
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
====================================================================================================
                                       coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------------
Intercept                          -11.8661      0.247    -48.132      0.000     -12.349     -11.383
color[T.yellow]                     -0.0924      0.102     -0.910      0.363      -0.292       0.107
payment[T.credit card]               8.2966      0.067    123.092      0.000       8.164       8.429
pickup_borough[T.Brooklyn]           1.9216      0.283      6.793      0.000       1.367       2.476
pickup_borough[T.Manhattan]          2.8871      0.261     11.064      0.000       2.376       3.399
pickup_borough[T.Queens]             3.3603      0.271     12.409      0.000       2.829       3.891
dropoff_borough[T.Brooklyn]          0.1444      0.245      0.588      0.556      -0.337       0.625
dropoff_borough[T.Manhattan]        -0.1799      0.223     -0.807      0.419      -0.617       0.257
dropoff_borough[T.Queens]            0.0172      0.238      0.072      0.942      -0.449       0.483
dropoff_borough[T.Staten Island]    -3.0338      1.500     -2.022      0.043      -5.975      -0.093
passengers_robust                    0.0425      0.021      1.981      0.048       0.000       0.085
distance_robust                     -0.0395      0.050     -0.786      0.432      -0.138       0.059
fare_robust                         -8.6338      0.137    -63.021      0.000      -8.902      -8.365
tolls_robust                        -1.0231      0.031    -33.061      0.000      -1.084      -0.962
total_robust                         8.6757      0.126     68.664      0.000       8.428       8.923
==============================================================================
Omnibus:                     2420.504   Durbin-Watson:                   1.967
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            16733.229
Skew:                          -1.663   Prob(JB):                         0.00
Kurtosis:                      10.230   Cond. No.                         170.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [15]:
df["dropoff_borough"].value_counts()

Manhattan        5172
Queens            534
Brooklyn          497
Bronx             136
Staten Island       2
Name: dropoff_borough, dtype: int64

`Nuestro set de datos está basado en los viajes en taxi realizado en la ciudad de Nueva York. Queremos predecir si las propinas (tip) se ven afectadas por alguna de las variables recogidas en el set, tales como distancia, pasajeros, peajes, zona de recogida y/o descarga, método de pago, ...`

`Tras realizar la normalización de nuestra variable respuesta (dependiente) y no conseguir obtener una muestra de datos normales, realizamos también una estandarización de las variables predictoras (independientes).`

`Ahora realizamos un test ANOVA para comprobar el efecto de nuestras variables predictoras (independientes) sobre nuestra variable respuesta (dependiente). En los resultados del test podemos observar que: ` 
- `el valor de R-squared es 0.873, al igual que Adj.R-squared, lo que nos indica que el 87.3 % de la variación de los datos es explicado por las variables predictoras.`
- `el coef, nos indica la tasa de cambio de la variable respuesta por cada unidad que aumenta la variable predictora. Si observamos el signo de este coeficiente podemos indicar si la relación es positiva o negativa entre la variable predictora y respuesta. Nos encontramos que tenemos una relación negativa en color[T.yellow], dropoff_borough[T.Manhattan], dropoff_borough[T.Staten Island], distance_robust, fare_robust y tolls_robust, y este signo lo que nos quiere decir es que cuando aumenta la variable predictora disminuye la variable respuesta. El resto tienen positiva, que indica, el aumento de la variable predictora hace que aumente el valor de la variable repuesta.`
- `la columna P>|t| nos indica el p-valor, que es la probabilidad de rechazar la hipótesis nula (H0) cuando es cierta. En este caso H0 es que la variable predictora no afecta a la variable respuesta, siendo nuestra hipótesis alternativa (H1) que si existe relación entre la variable predictora y la variable respuesta. Si es < 0.05 rechazamos H0 y aceptamos la hipótesis alternativa. Al observar los resultados del test ANOVA vemos que las variables predictoras que no afectan a la variable respuesta son 'color' y 'distance_robust' (p-valor > 0.05). En el caso de 'dropoff_borough' tenemos que una de sus categorías si afecta a la variable respuesta a pesar de que el resto de categorías no, por lo que no podemos rechazar esta variable.`


**Happy coding** 🤯📊